c++/mpi/window.hpp

```diff
@@ -0,0 +1,183 @@
+      }
+    }
+
+    /// Completes an RMA access epoch started by a call to lock()
+    void unlock(int rank = -1) const noexcept {
+      if (rank < 0) {
+        MPI_Win_unlock_all(win);
+      } else {
+        MPI_Win_unlock(rank, win);
+      }
+    }
+
+    /// Load data from a remote memory window.
+    template <typename TargetType = BaseType, typename OriginType>
+    std::enable_if_t<has_mpi_type<OriginType> && has_mpi_type<TargetType>, void>
+    get(OriginType *origin_addr, int origin_count, int target_rank, MPI_Aint target_disp = 0, int target_count = -1) const noexcept {
+      MPI_Datatype origin_datatype = mpi_type<OriginType>::get();
+      MPI_Datatype target_datatype = mpi_type<TargetType>::get();
+      int target_count_ = target_count < 0 ? origin_count : target_count;
+      MPI_Get(origin_addr, origin_count, origin_datatype, target_rank, target_disp, target_count_, target_datatype, win);
+    };
+
+    /// Store data to a remote memory window.
+    template <typename TargetType = BaseType, typename OriginType>
+    std::enable_if_t<has_mpi_type<OriginType> && has_mpi_type<TargetType>, void>
+    put(OriginType *origin_addr, int origin_count, int target_rank, MPI_Aint target_disp = 0, int target_count = -1) const noexcept {
+      MPI_Datatype origin_datatype = mpi_type<OriginType>::get();
+      MPI_Datatype target_datatype = mpi_type<TargetType>::get();
+      int target_count_ = target_count < 0 ? origin_count : target_count;
+      MPI_Put(origin_addr, origin_count, origin_datatype, target_rank, target_disp, target_count_, target_datatype, win);
+    };
+
+    /// Accumulate data into target process through remote memory access.
+    template <typename TargetType = BaseType, typename OriginType>
+    std::enable_if_t<has_mpi_type<OriginType> && has_mpi_type<TargetType>, void>
+    accumulate(OriginType const *origin_addr, int origin_count, int target_rank, MPI_Aint target_disp = 0, int target_count = -1, MPI_Op op = MPI_SUM) const noexcept {
+      MPI_Datatype origin_datatype = mpi_type<OriginType>::get();
+      MPI_Datatype target_datatype = mpi_type<TargetType>::get();
+      int target_count_ = target_count < 0 ? origin_count : target_count;
+      MPI_Accumulate(origin_addr, origin_count, origin_datatype, target_rank, target_disp, target_count_, target_datatype, op, win);
+    }
+
+    /// Obtains the value of a window attribute.
+    void* get_attr(int win_keyval) const noexcept {
+      int flag;
+      void *attribute_val;
+      MPI_Win_get_attr(win, win_keyval, &attribute_val, &flag);
+      assert(flag);
+      return attribute_val;
+    }
+
+    // Expose some commonly used attributes
+    BaseType* base() const noexcept { return static_cast<BaseType*>(get_attr(MPI_WIN_BASE)); }
+    MPI_Aint size() const noexcept { return *static_cast<MPI_Aint*>(get_attr(MPI_WIN_SIZE)); }
+    int disp_unit() const noexcept { return *static_cast<int*>(get_attr(MPI_WIN_DISP_UNIT)); }
+  };
```

Comment:
Can these function calls be turned into members that are initialized at construction?